File: app/src/main/java/md/fusionworks/lifehack/ui/activity/BaseActivity.java
package md.fusionworks.lifehack.ui.activity;

import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;

import butterknife.Bind;
import butterknife.ButterKnife;
import md.fusionworks.lifehack.R;
import md.fusionworks.lifehack.ui.listener.NotificationToastActionListener;
import md.fusionworks.lifehack.ui.navigator.Navigator;
import md.fusionworks.lifehack.ui.widget.LoadingDialog;
import md.fusionworks.lifehack.util.Constant;

/**
 * Created by ungvas on 10/15/15.
 */
public class BaseActivity extends AppCompatActivity {

  @Nullable @Bind(R.id.coordinatorLayout) CoordinatorLayout coordinatorLayout;

  private Navigator navigator;
  private LoadingDialog loadingDialog;

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    navigator = new Navigator();
    loadingDialog = new LoadingDialog(this);
  }

  @Override public void setContentView(@LayoutRes int layoutResID) {
    super.setContentView(layoutResID);
    ButterKnife.bind(this);
  }

  protected void addFragment(int containerViewId, Fragment fragment) {
    android.support.v4.app.FragmentTransaction fragmentTransaction =
        this.getSupportFragmentManager().beginTransaction();
    fragmentTransaction.replace(containerViewId, fragment, null);
    fragmentTransaction.commit();
  }

  protected void replaceFragment(int containerViewId, Fragment fragment) {
    android.support.v4.app.FragmentTransaction fragmentTransaction =
        this.getSupportFragmentManager().beginTransaction();
    fragmentTransaction.replace(containerViewId, fragment, null);
    fragmentTransaction.commit();
  }

  public void showNotificationToast(int type, String message) {
    if (coordinatorLayout != null) {
      coordinatorLayout.postDelayed(() -> {
        Snackbar snackbar = Snackbar.make(coordinatorLayout, message, Snackbar.LENGTH_SHORT);
        if (type == Constant.NOTIFICATION_TOAST_ERROR) {
          snackbar.getView().setBackgroundColor(Color.RED);
        }
        snackbar.show();
      }, Constant.NOTIFICATION_TOAST_SHOW_DELAY);
    }
  }

  public void showNotificationToast(int type, String message,
      NotificationToastActionListener notificationToastActionListener) {
    if (coordinatorLayout != null) {
      coordinatorLayout.postDelayed(() -> {
        Snackbar snackbar = Snackbar.make(coordinatorLayout, message, Snackbar.LENGTH_INDEFINITE);
        if (type == Constant.NOTIFICATION_TOAST_ERROR) {
          snackbar.getView().setBackgroundColor(Color.RED);
        }
        snackbar.setAction(R.string.action_retry, v -> {
          if (notificationToastActionListener != null) notificationToastActionListener.onClick();
        });
        snackbar.show();
      }, Constant.NOTIFICATION_TOAST_SHOW_DELAY);
    }
  }

  public Navigator getNavigator() {
    return navigator;
  }

  public void showLoadingDialog() {
    if (this.loadingDialog != null) {
      this.loadingDialog.show();
    }
  }

  public void hideLoadingDialog() {
    if (this.loadingDialog != null) {
      this.loadingDialog.dismiss();
    }
  }
}

File: app/src/main/java/md/fusionworks/lifehack/view/activity/BaseActivity.kt
package md.fusionworks.lifehack.view.activity

import android.graphics.Color
import android.os.Bundle
import android.support.design.widget.CoordinatorLayout
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import com.trello.rxlifecycle.components.support.RxAppCompatActivity
import md.fusionworks.lifehack.R
import md.fusionworks.lifehack.application.AppComponent
import md.fusionworks.lifehack.application.LifehackApp
import md.fusionworks.lifehack.navigator.Navigator
import md.fusionworks.lifehack.util.Constant
import md.fusionworks.lifehack.view.widget.LoadingDialog
import org.jetbrains.anko.find
import rx.subscriptions.CompositeSubscription

/**
 * Created by ungvas on 10/15/15.
 */
open class BaseActivity : RxAppCompatActivity() {

  val coordinatorLayout by lazy { find<CoordinatorLayout>(R.id.coordinatorLayout) }

  lateinit var navigator: Navigator
  private lateinit var loadingDialog: LoadingDialog
  protected lateinit var loadingDialogCancelSubscription: CompositeSubscription

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    navigator = Navigator()
    loadingDialog = LoadingDialog(this) {
      loadingDialogCancelSubscription.clear()
      onLoadingDialogCanceled()
    }
    loadingDialogCancelSubscription = CompositeSubscription()
  }

  override fun onResume() {
    super.onResume()
    listenForEvents()
  }

  protected fun addFragment(containerViewId: Int, fragment: Fragment) {
    val fragmentTransaction = this.supportFragmentManager.beginTransaction()
    fragmentTransaction.replace(containerViewId, fragment, null)
    fragmentTransaction.commit()
  }

  fun showNotificationToast(type: Int, message: String) {
    if (coordinatorLayout != null) {
      coordinatorLayout.postDelayed({
        val snackbar = Snackbar.make(coordinatorLayout, message, Snackbar.LENGTH_SHORT)
        if (type == Constant.NOTIFICATION_TOAST_ERROR) {
          snackbar.view.setBackgroundColor(Color.RED)
        }
        snackbar.show()
      }, Constant.NOTIFICATION_TOAST_SHOW_DELAY.toLong())
    }
  }

  fun showNotificationToast(type: Int, message: String, onRetry: () -> Unit) {
    if (coordinatorLayout != null) {
      coordinatorLayout.postDelayed({
        val snackbar = Snackbar.make(coordinatorLayout, message, Snackbar.LENGTH_INDEFINITE)
        if (type == Constant.NOTIFICATION_TOAST_ERROR) {
          snackbar.view.setBackgroundColor(Color.RED)
        }
        snackbar.setAction(R.string.retry) { v -> onRetry() }
        snackbar.show()
      }, Constant.NOTIFICATION_TOAST_SHOW_DELAY.toLong())
    }
  }

  fun showLoadingDialog() {
    if (this.loadingDialog != null) {
      this.loadingDialog.show()
    }
  }

  val appComponent: AppComponent
    get() = LifehackApp.component

  fun hideLoadingDialog() {
    if (this.loadingDialog != null) {
      this.loadingDialog.dismiss()
    }
  }

  protected open fun listenForEvents() {
  }

  protected fun onLoadingDialogCanceled() {
  }

  protected open fun initializeDIComponent() {
  }
}
